### PR TITLE
feat(FEC-9341): refactor base component to HoC

### PR DESCRIPTION
**Pull request by OrenMe**
### Description of the Changes

* added PlayerProvider and withPlayer components that expose player as prop to components
* added EventDispatcher and withEventDispatcher components that inject EventDispatcher actions from context to props of a component
* added withLogger component that inject logger as props of a component with a supplied namespace
* added withEventManager component that inject event manager as props of a component that require to manage event listeners
* fixed withAnimtation HoC signature to be inline with rest of HoCs
*moved `shouldRender` to be internal in component - this is cause HoCs don't expose static methods and in general this is not the best way to do it if we need to be able to dynamically add and remove components from bundle
* updated all indices of the player to use the new HoCs
* added doc

### CheckLists

- [x] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] test are passing in local environment 
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [x] Docs have been updated

### CODE REVIEW

File: src/components/logo/logo.js [hunk 1]
```diff
@@ -20,31 +21,22 @@ const mapStateToProps = state => ({
...
-    return !(Object.keys(componentConfig).length === 0 && componentConfig.constructor === Object);
-  }
-  /**
-   * Creates an instance of Logo.
-   * @param {Object} obj obj
-   * @memberof Logo
    */
-  constructor(obj: Object) {
-    super({name: COMPONENT_NAME, player: obj.player});
+  _shouldRender(): boolean {
```
Comment (esakal) [inline src/components/logo/logo.js]: is it here on purpose?

File: src/components/seekbar/seekbar.js [hunk 2]
```diff
@@ -25,6 +26,7 @@ const COMPONENT_NAME = 'SeekBar';
   mapStateToProps,
   bindActions(actions)
 )
+@withPlayer
```
Comment (esakal) [inline src/components/seekbar/seekbar.js]: although it doesn't inherit from baseComponent you might consider with logger here

File: src/components/time-display-ads-container/time-display-ads-container.js [hunk 3]
```diff
@@ -21,17 +20,9 @@ const COMPONENT_NAME = 'TimeDisplayAdsContainer';
...
-class TimeDisplayAdsContainer extends BaseComponent {
-  /**
-   * Creates an instance of TimeDisplayAdsContainer.
-   * @memberof TimeDisplayAdsContainer
-   */
-  constructor() {
-    super({name: COMPONENT_NAME});
-  }
-
+class TimeDisplayAdsContainer extends Component {
```
Comment (esakal) [inline src/components/time-display-ads-container/time-display-ads-container.js]: logger?

File: src/components/copy-button/copy-button.js [hunk 4]
```diff
@@ -20,9 +19,9 @@ const COMPONENT_NAME = 'CopyButton';
  *
  * @class CopyButton
  * @example <CopyButton/>
- * @extends {BaseComponent}
+ * @extends {Component}
  */
-class CopyButton extends BaseComponent {
+class CopyButton extends Component {
```
[the remark this answers is not in the document]
Comment (OrenMe) [reply src/components/copy-button/copy-button.js]: Seems like this was never meant to be BaseComponent at all so I removed it and no logger for it as well

File: src/components/event-dispatcher/event-dispatcher-provider.js [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+ * @extends {Component}
+ */
+class EventDispatcherProvider extends Component {
+  /**
+   * create context player
+   * @returns {void}
+   */
+  getChildContext() {
+    return {
+      notifyClick: (payload: Object) => onClickableComponentsHandler(this.props.store, payload, this.props.player),
```
[the remark this answers is not in the document]
Comment (OrenMe) [reply src/components/event-dispatcher/event-dispatcher-provider.js]: done

File: src/components/time-display-playback-container/time-display-playback-container.js [hunk 6]
```diff
@@ -17,22 +17,15 @@ const mapStateToProps = state => ({
 const COMPONENT_NAME = 'TimeDisplayPlaybackContainer';
 
 @connect(mapStateToProps)
+@withLogger(COMPONENT_NAME)
 /**
  * TimeDisplayPlaybackContainer component
  *
  * @class TimeDisplayPlaybackContainer
  * @example <TimeDisplayPlaybackContainer format='currentTime / duration' />
  * @extends {BaseComponent}
```
Comment (yairans) [inline src/components/time-display-playback-container/time-display-playback-container.js]:
```suggestion
  * @extends {Component}
```

File: src/components/event-dispatcher/with-event-dispatcher.js [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+ * @returns {Component} - the wrapped component
+ */
+export const withEventDispatcher: Function = (name: string) => (WrappedComponent: Component): typeof Component =>
+  class EventDispatcherComponent extends Component {
+    /**
+     * Notify that a clickable component has been clicked.
+     * @param {any} payload - Optional payload.
+     * @returns {void}
+     *
+     * @memberof BaseComponent
```
Comment (yairans) [inline src/components/event-dispatcher/with-event-dispatcher.js]:
```suggestion
     * @memberof EventDispatcherComponent
```

File: src/components/settings/settings.js [hunk 8]
```diff
@@ -232,7 +232,7 @@ class Settings extends BaseComponent {
                 />
               </Localizer>
             )}
-            {props.isLive || speedOptions.length <= 1 ? (
+            {isLive || speedOptions.length <= 1 ? (
```
Comment (yairans) [inline src/components/settings/settings.js]: Do the same above (line 207)

File: src/components/share-overlay/share-overlay.js [hunk 9]
```diff
@@ -138,34 +138,26 @@ const COMPONENT_NAME = 'ShareOverlay';
   mapStateToProps,
   bindActions(actions)
 )
+@withLogger(COMPONENT_NAME)
```
Comment (yairans) [inline src/components/share-overlay/share-overlay.js]: @withPlayer

File: src/components/share-overlay/share-overlay.js [hunk 10]
```diff
@@ -1,20 +1,20 @@
...
 import {connect} from 'preact-redux';
 import {bindActions} from '../../utils/bind-actions';
 import {actions} from '../../reducers/share';
 import {toHHMMSS, toSecondsFromHHMMSS} from '../../utils/time-format';
-import BaseComponent from '../base';
 import {Overlay} from '../overlay';
 import {default as Icon, IconType} from '../icon';
 import style from '../../styles/style.scss';
 import {CopyButton} from '../copy-button/copy-button';
+import {withLogger} from 'components/logger';
```
Comment (yairans) [inline src/components/share-overlay/share-overlay.js]: import {withPlayer} from '../player';

File: src/components/settings/settings.js (hunk 8 above)
Comment (OrenMe) [reply src/components/settings/settings.js]: done

File: src/components/share-overlay/share-overlay.js (hunk 10 above)
Comment (OrenMe) [reply src/components/share-overlay/share-overlay.js]: share overlay get its player as prop already from before, from the share component

File: src/components/logo/logo.js (hunk 1 above)
Comment (OrenMe) [reply src/components/logo/logo.js]: so the shouldRender was there, see a few rows above :-)

**Pull request merged**